app/core/api_client.py

```diff
@@ -41,78 +49,112 @@ async def _make_request(
+            )
             return {
                 "status_code": 503,  # Service Unavailable
                 "data": None,
-                "error": f"Connection error: {str(error)}"
+                "error": f"Connection error: {str(error)}",
             }
         except httpx.HTTPStatusError as error:
             # Log and handle HTTP errors (non-2xx responses)
-            logger.error(f"Error response {error.response.status_code} "
-                         f"while requesting {error.request.url!r}.")
+            logger.error(
+                f"Error response {error.response.status_code} "
+                f"while requesting {error.request.url!r}."
+            )
             return {
                 "status_code": error.response.status_code,
                 "data": error.response.json(),
-                "error": f"HTTP error: {error.response.status_code} - {error.response.text}"
+                "error": f"HTTP error: {error.response.status_code} - {error.response.text}",
             }
         except Exception as error:
             # Catch any other unexpected errors
             logger.error(f"An unexpected error occurred: {str(error)}")
             return {
                 "status_code": 500,  # Internal Server Error
                 "data": None,
-                "error": f"Unexpected error: {str(error)}"
+                "error": f"Unexpected error: {str(error)}",
             }
 
-    async def get(self, endpoint: str, headers: dict[str, Any] | None = None,
-                  params: dict[str, Any] | None = None) -> Any:
-        response = await self._make_request("GET", endpoint, params=params, headers=headers)
+    async def get(
+        self,
+        endpoint: str,
+        headers: dict[str, Any] | None = None,
+        params: dict[str, Any] | None = None,
+    ) -> Any:
+        response = await self._make_request(
+            "GET", endpoint, params=params, headers=headers
+        )
         return response
 
-    async def post(self, endpoint: str, headers: dict[str, Any] | None = None,
-                   data: dict[str, Any] | None = None) -> Any:
-        response = await self._make_request("POST", endpoint, data=data, headers=headers)
+    async def post(
+        self,
+        endpoint: str,
+        headers: dict[str, Any] | None = None,
+        data: dict[str, Any] | None = None,
+    ) -> Any:
+        response = await self._make_request(
+            "POST", endpoint, data=data, headers=headers
+        )
         return response
 
-    async def put(self, endpoint: str, headers: dict[str, Any] | None = None,
-                  data: dict[str, Any] | None = None) -> Any:
+    async def put(
+        self,
+        endpoint: str,
+        headers: dict[str, Any] | None = None,
+        data: dict[str, Any] | None = None,
+    ) -> Any:
         response = await self._make_request("PUT", endpoint, data=data, headers=headers)
         return response
```

Comment:
yes, it's an alternative solution.

Reply:
That's a good point about typing, didn't think of that, however:

1. It seems that recent versions of `mypy` deals with partials well after some exploration: https://mypy-play.net/?mypy=latest&python=3.12&gist=41e197646be3ad54f4b854c411254ac2 Other type checkers may not but they are not really relevant as we only have mypy in the CI pipeline
2. I suspect these incompatibilities are more likely to be related to Python versions as advanced type hints like [Concatenate](https://docs.python.org/3/library/typing.html#typing.Concatenate) are relatively recent additions to the language, so it's very likely this was a real problem before but not anymore :)
3. In general, our typing checks are not super strict, we don't have strict mode enabled and there are a lot of minor mistakes (which are still enough to throw off type checkers and can return misleading or not useful autocompletion results). For example -- the pydantic schemas of fastapi's dependancies technically should be using [Annotated](https://docs.python.org/3/library/typing.html#typing.Annotated). On one hand you're right about all the typing benefits and I completely agree but on the other hand it sometimes takes so much time (in development, during review or blocking the pipeline if we enforce it with automation) that the benefits may not be worth it. I can't form a strong opinion either way as they have pros and cons but I'm curious what you guys thing? Because at the moment I don't know for example if it's useful to spend time on that during review